Python/ceval.c

```diff
@@ -2213,24 +2213,17 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
         case TARGET(YIELD_FROM): {
             PyObject *v = POP();
             PyObject *receiver = TOP();
-            int is_gen_or_coro = PyGen_CheckExact(receiver) || PyCoro_CheckExact(receiver);
-            int gen_status;
-            if (tstate->c_tracefunc == NULL && is_gen_or_coro) {
-                gen_status = PyGen_Send((PyGenObject *)receiver, v, &retval);
+            PySendResult gen_status;
+            if (tstate->c_tracefunc == NULL) {
```

Comment:
Can we fold the tracing into the `genobject`/`PyIter_Send` inmplementation to avoid code duplication?

Reply:
That would also make ceval a tad smaller which is always a good thing.

Reply:
I do not think it is feasible.

Reply:
Can you elaborate?

Reply:
There is a call of the trace function in between calls of `__next__`/`send` and `_PyGen_FetchStopIterationValue`. It cannot be called after `PyIter_Send` because the original exception (which can be a subclass of `StopIteration` or `StopIteration` with additional attributes) is already silenced in `PyIter_Send`. Therefore we should pass the trace function to `PyIter_Send`, and this would complicate the API.

We can introduce a private function `_PyIter_SendWithTrace` which accepts also the trace function argument. But would not it make the code more complicated? In any case it is better to left to a separate PR or separate issue.

Reply:
I was thinking about exporting the `call_exc_trace` function and calling it directly from `genobject.c`; the trace function can easily be read from the thread state.